File: bin/pythiaRun.cc
// main42.cc is a part of the PYTHIA event generator.
// Copyright (C) 2013 Torbjorn Sjostrand.
// PYTHIA is licenced under the GNU GPL version 2, see COPYING for details.
// Please respect the MCnet Guidelines, see GUIDELINES for details.

// It illustrates how Les Houches Event File input can be used in PYTHIA
// and events written to HepMC format


#include "Pythia8/Pythia.h"
#include "Pythia8Plugins/HepMC2.h"
#include "HepMC/GenEvent.h"   
#include "HepMC/IO_GenEvent.h"

using namespace Pythia8; 
using namespace std;

int main(int argc, char* argv[]) {

  // Check that correct number of command-line arguments
  if (argc < 2) {
    cerr << " Unexpected number of command-line arguments." << endl
         << " Usage:" << endl
         << "    pythiaRun <card-file> <output-HepMCfile>" << endl;
    return 1;
  }

  // Check that the provided input name corresponds to an existing file.
  ifstream is(argv[1]);  
  if (!is) {
    cerr << " Command-line file " << argv[1] << " was not found. \n"
         << " Program stopped! " << endl;
    return 1;
  }
  // Confirm that external files will be used for input and output.
  cout << "\n >>> PYTHIA settings will be read from file " << argv[1] << endl;


  string HepMCfile(".HepMCfile");
  if (argc > 2)
    {
      HepMCfile = string(argv[2]);
    cout  << " HepMC events will be written to file " 
	  << HepMCfile << " <<< \n" << endl;
    }

  // Interface for conversion from Pythia8::Event to HepMC event. 
  HepMC::Pythia8ToHepMC ToHepMC;
  // Specify file where HepMC events will be stored.
  HepMC::IO_GenEvent ascii_io(HepMCfile.c_str(), std::ios::out);
 
  // Generator. 
  Pythia pythia;

  // Read in commands from external file.
  pythia.readFile(argv[1]);    

  // Extract settings to be used in the main program.
  int    nEvent    = pythia.mode("Main:numberOfEvents");
  int    nAbort    = pythia.mode("Main:timesAllowErrors");
 
  // Initialization.
  pythia.init();

  // Begin event loop.
  int iAbort = 0; 
  for (int iEvent = 0; iEvent < nEvent; ++iEvent) {

    // Generate event. 
    if (!pythia.next()) {

      // If failure because reached end of file then exit event loop.
      if (pythia.info.atEndOfFile()) {
        cout << " Aborted since reached end of Les Houches Event File\n"; 
        break; 
      }

      // First few failures write off as "acceptable" errors, then quit.
      if (++iAbort < nAbort) continue;
      cout << " Event generation aborted prematurely, owing to error!\n"; 
      break;
    }

    if (argc > 2)
      {
	// Construct new empty HepMC event and fill it.
	// Units will be as chosen for HepMC build, but can be changed
	// by arguments, e.g. GenEvt( HepMC::Units::GEV, HepMC::Units::MM)  
	HepMC::GenEvent* hepmcevt = new HepMC::GenEvent();
	ToHepMC.fill_next_event( pythia, hepmcevt );
	
	// Write the HepMC event to file. Done with it.
	ascii_io << hepmcevt;
	delete hepmcevt;
      }

  // End of event loop. Statistics. 
  }
  pythia.stat();

  // Done.
  return 0;
}

File: bin/pythiaRun.py
#!/usr/bin/env python
#-----------------------------------------------------------------------------
# File: pythiaRun.py
#-----------------------------------------------------------------------------
import os, sys
import pythia8 as p8
import pyHepMC3 as hp
# linked HepMC3/python/test to lib/python3.9/site-packages
from Pythia8ToHepMC3 import Pythia8ToHepMC3
#-----------------------------------------------------------------------------
WriterAscii = hp.HepMC3.WriterAscii
ReaderAscii = hp.HepMC3.ReaderAscii
GenEvent    = hp.HepMC3.GenEvent
Units       = hp.HepMC3.Units
Print       = hp.HepMC3.Print
#-----------------------------------------------------------------------------
def get_filenames():
    argv = sys.argv[1:]
    argc = len(argv)
    if argc < 1:
        sys.exit('''
        Usage:
        02_study.py control-file-name
        ''')
    inpfilename = argv[0]
    if not os.path.exists(inpfilename):
        sys.exit('''
        file %s NOT found!
        ''' % inpfilename)
    outfilename = inpfilename.replace('.txt', '.hepmc3')

    print('control file: %s' % inpfilename)
    print('output  file: %s' % outfilename)
    
    return inpfilename, outfilename
#-----------------------------------------------------------------------------
def main():
    
    inpfilename, outfilename = get_filenames()

    toHepMC3 = Pythia8ToHepMC3()
    writer   = WriterAscii(outfilename)

    # Read Pythia instructions from control file
    pythia   = p8.Pythia()
    pythia.readFile(inpfilename)

    nEvent   = pythia.mode("Main:numberOfEvents")
    nAbort   = pythia.mode("Main:timesAllowErrors")

    print("number of events to generate         %d" % nEvent)
    print("maximum number of events bad events  %d" % nAbort)
    
    # Initialize Pythia
    pythia.init()

    Abort = 0
    for event_number in range(nEvent):

        # generate event
        if pythia.next():

            # write ouut event in HepMC3 format
            hepmcevt = GenEvent()
            toHepMC3.fill_next_event1(pythia, hepmcevt, event_number)
            writer.write_event(hepmcevt)
            
        else:
            # we could be reading from an LHE file, so check
            # for end of file.
            if pythia.info.atEndOfFile():
                print("** reached end of LHE file")
                break
            
            Abort += 1
            if Abort < nAbort:
                print("** abort event")
                break

    pythia.stat()
#-----------------------------------------------------------------------------
try:
    main()
except KeyboardInterrupt:
    print('\nbye!')
    

